Add explicit dynamic fusion arg

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #72077
* #72036
* #72038
* #72040
* #72032
* #71651
* #71650
* __->__ #71173

Differential Revision: [D33536222](https://our.internmc.facebook.com/intern/diff/D33536222)

Commit:
[WIP] add explicit dynamic fusion arg

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33536222).

Commit:
Update on "[WIP] add explicit dynamic fusion arg"


Differential Revision: [D33536222](https://our.internmc.facebook.com/intern/diff/D33536222)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33536222).

Commit:
Update on "[WIP] add explicit dynamic fusion arg"


Differential Revision: [D33536222](https://our.internmc.facebook.com/intern/diff/D33536222)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33536222).

Commit:
Update on "[WIP] add explicit dynamic fusion arg"


Differential Revision: [D33536222](https://our.internmc.facebook.com/intern/diff/D33536222)

[ghstack-poisoned]

navahgar commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -1236,7 +1249,8 @@ void FuseTensorExprs(
 
 Operation createTensorExprOp(const Node* node) {
-  if (!tensorExprDynamicShapeFusionEnabled()) {
+  bool dynamic_shape_fusion_node = node->hasAttribute(attr::striding_inputs_desc);
```
Is every `TensorExprGroup` node with dynamic shapes guaranteed to have `attr::striding_inputs_desc`? What if there are no dynamic strides?

eellison replied:
yes always guaranteed. If there are no dynamic strides then there will be `attr::striding_inputs_desc` with TENSOR_CONT description

navahgar commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -1210,23 +1216,30 @@ class TensorExprFuser {
   }
 
+  if (add_composed_op) {
+    TORCH_INTERNAL_ASSERT(fuse_to_dynamic_shapes, "Fusing static shapes with composed op NYI");
```
Why is this an error?

eellison replied:
The `add_composed_op` path always fuses to dynamic shapes right now, we didn't implement composed op + static shapes (bc there is no use case for it)

navahgar reviewed (commented):
Mostly LGTM. Couple of minor comments.

navahgar reviewed (approved):
LGTM

Commit:
Update on "Add explicit dynamic fusion arg"


Differential Revision: [D33536222](https://our.internmc.facebook.com/intern/diff/D33536222)

[ghstack-poisoned]

Commit:
Update on "Add explicit dynamic fusion arg"


Differential Revision: [D33536222](https://our.internmc.facebook.com/intern/diff/D33536222)

[ghstack-poisoned]

Commit:
Update on "Add explicit dynamic fusion arg"


Differential Revision: [D33536222](https://our.internmc.facebook.com/intern/diff/D33536222)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33536222).

Commit:
Update on "Add explicit dynamic fusion arg"


Differential Revision: [D33536222](https://our.internmc.facebook.com/intern/diff/D33536222)

[ghstack-poisoned]

Commit:
Update on "Add explicit dynamic fusion arg"


Differential Revision: [D33536222](https://our.internmc.facebook.com/intern/diff/D33536222)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33536222).

Commit:
Update on "Add explicit dynamic fusion arg"


Differential Revision: [D33536222](https://our.internmc.facebook.com/intern/diff/D33536222)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33536222).

Commit:
Update on "Add explicit dynamic fusion arg"


Differential Revision: [D33536222](https://our.internmc.facebook.com/intern/diff/D33536222)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33536222).

xsacha commented:
This commit appears to have broken some of my models at specific input sizes.
See #72173
